### PR TITLE
Preliminary Intel VT-d IOMMU support

### PR DESCRIPTION
**Pull request by NicholasLindsay**
# Overview

Adds preliminary driver support for an Intel IOMMU implementing Intel VT-d (Intel Virtualization for Directed I/O). Currently all we do is detect the IOMMU, and if detected, print out some useful information. This sets the groundwork for the rest of the IOMMU implementation.

This preliminary driver performs the following:
* Initialize the kernel IOMMU object if it is detected in the ACPI DMAR table (see `acpi::init()` and `iommu::early_init()`)
   * Map the IOMMU's memory-mapped registers into virtual memory
* Output the interpreted contents of the following IOMMU registers to the log:
  * `Version Register`
  * `Capability Register`
  * `Extended Capability Register`
  * `Global Status Register`
* Further initialization will be performed by calling `iommu::init()` from  `device_manager::init()` - although currently `iommu::init()` doesn't do anything useful.

Additionally, the driver provides a mechanism for sending commands to the IOMMU `Global Command Register` (and waiting until the command is complete).

# Testing

To test IOMMU support, pass `IOMMU=yes` when invoking the Makefile QEMU target (`orun` etc). Details about the IOMMU will be printed to the log.

Please test with and without `IOMMU=yes`.

# Outstanding issues:
* Printing to the log the interpreted contents of the `Capability Register` and `Extended Capability Register` produces multiple lines, of which only the first line has the log prefix (e.g. `[I] ...`)
* Only a single IOMMU is supported. Support for multiple IOMMUs can be added in the future.
* Further initialization in `device_manager::init()` currently doesn't do anything useful but is there because I anticipate it will be needed in the future.

### CODE REVIEW

**Comment by NicholasLindsay** [comment]
This PR is very much a work-in-progress and there will be multiple iterations until IOMMU support is fully ready. I am publishing this PR early in the hope that I can get some early feedback on coding style (this is my first PR for Theseus with actual code).

File: kernel/iommu/src/regs.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        writeln!(f, "PSI:     {}", self.psi())?;
+        writeln!(f, "NFR:     {}", self.nfr())?;
+        writeln!(f, "MAMV:    {}", self.mamv())?;
+        writeln!(f, "DWD:     {}", self.dwd())?;
+        writeln!(f, "DRD:     {}", self.drd())?;
+        writeln!(f, "FL1GP:   {}", self.fl1gp())?;
+        writeln!(f, "PI:      {}", self.pi())?;
+        writeln!(f, "FL5LP:   {}", self.fl5lp())?;
+        writeln!(f, "ESIRTPS: {}", self.esirtps())?;
+        writeln!(f, "ESRTPS:  {}", self.esrtps())?;
```
Comment (kevinaboos) [inline kernel/iommu/src/regs.rs]: for this, check out [`debug_struct`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct)

**Review by kevinaboos — COMMENTED** [review]
The code style so far looks quite good! I don't see any glaring problems. I left a few small comments, nothing major. I'll let you guide the progress of this PR; feel free to reach out to me on Slack as well. 

There's probably a better way to represent the `Capability` and `ExtendedCapability` structs, but what you've done thus far isn't necessarily wrong or bad. I'd need to see more about how those types are used in order to suggest something better, but you could always look at `bitfield` or `bitflags` for inspiration. I will say that if you do stick with this approach, it's probably clearer and more efficient to test a specific bit or a range of bits rather than right-shift the values by that bit index and then test against `0x1`, but I'm not sure if it actually compiles down to something much different. 

My primary question is whether it's necessary to split IOMMU initialization into `early_init()` and regular `init()`. That's only really needed when we have to break cyclic dependencies, e.g., two subsystems that partially depend on each other. If not, then i'd just stick with one regular `init()` function for simplicity's sake.

**Comment by NicholasLindsay** [comment]
Is it okay to force-push changes after a local rebase?

**Comment by kevinaboos** [comment]
> Is it okay to force-push changes after a local rebase?

Sure, that's no problem on your own private local branch.

**Comment by NicholasLindsay** [comment]
> > Is it okay to force-push changes after a local rebase?
> 
> Sure, that's no problem on your own private local branch.

For more safety better to use `--force-with-lease`, even on your own private local branch.

**Comment by NicholasLindsay** [comment]
And is it okay to use `git commit --amend` to update PRs in response to comments? Or do you prefer separate commits (perhaps with a squash at the end)?

**Comment by kevinaboos** [comment]
> And is it okay to use `git commit --amend` to update PRs in response to comments? Or do you prefer separate commits (perhaps with a squash at the end)?

Oh yeah, don't worry about commit history. Pushing new individual commits is much easier for me to track. I'll squash everything later when it's time to merge anyway. 

It sounds like you intend to make many more commits; what else are you planning for this PR?

**Comment by NicholasLindsay** [comment]
I agree with your advice; there is no need for separate `early_init()` and`init()` functions - at least not currently. I have merged these two functions together into a single `init()` function which is called from `acpi::init()`.

**Comment by NicholasLindsay** [comment]
I haven't decided exactly how I want to structure the PRs yet. My initial thought was that the push into the mainline could happen once the IOMMU is up and running and is providing isolation between devices and the CPU.

**Comment by NicholasLindsay** [comment]
> I haven't decided exactly how I want to structure the PRs yet. My initial thought was that the push into the mainline could happen once the IOMMU is up and running and is providing isolation between devices and the CPU.

There are likely to be many commits between now and then due to the general complexity of IOMMUs. If you prefer, I can provide smaller, more incremental PRs.

**Comment by NicholasLindsay** [comment]
> The code style so far looks quite good! I don't see any glaring problems. I left a few small comments, nothing major. I'll let you guide the progress of this PR; feel free to reach out to me on Slack as well.
> 
> There's probably a better way to represent the `Capability` and `ExtendedCapability` structs, but what you've done thus far isn't necessarily wrong or bad. I'd need to see more about how those types are used in order to suggest something better, but you could always look at `bitfield` or `bitflags` for inspiration. I will say that if you do stick with this approach, it's probably clearer and more efficient to test a specific bit or a range of bits rather than right-shift the values by that bit index and then test against `0x1`, but I'm not sure if it actually compiles down to something much different.
> 
> My primary question is whether it's necessary to split IOMMU initialization into `early_init()` and regular `init()`. That's only really needed when we have to break cyclic dependencies, e.g., two subsystems that partially depend on each other. If not, then i'd just stick with one regular `init()` function for simplicity's sake.

Perhaps something like this would be clearer for checking bit values: `(self.0) & (1 << 63) != 0`? It's slightly more convoluted than the C version since in Rust there is no implicit conversion from `u64` to `bool`.

**Comment by NicholasLindsay** [comment]
> The code style so far looks quite good! I don't see any glaring problems. I left a few small comments, nothing major. I'll let you guide the progress of this PR; feel free to reach out to me on Slack as well.
> 
> There's probably a better way to represent the `Capability` and `ExtendedCapability` structs, but what you've done thus far isn't necessarily wrong or bad. I'd need to see more about how those types are used in order to suggest something better, but you could always look at `bitfield` or `bitflags` for inspiration. I will say that if you do stick with this approach, it's probably clearer and more efficient to test a specific bit or a range of bits rather than right-shift the values by that bit index and then test against `0x1`, but I'm not sure if it actually compiles down to something much different.
> 
> My primary question is whether it's necessary to split IOMMU initialization into `early_init()` and regular `init()`. That's only really needed when we have to break cyclic dependencies, e.g., two subsystems that partially depend on each other. If not, then i'd just stick with one regular `init()` function for simplicity's sake.

`bitflags` won't work since there are some fields than span multiple bits. And I had a look at `bitfield` but I thought it looked rather cumbersome. Additionally there are some IOMMU register fields which when read require further arithmetic (e.g. `fn nfr(&self)     -> u64  { ((self.0 >> 40) & 0xff) + 1 }`), which cannot be done cleanly with `bitfield`.

I have used the `bitflags!` macro for `GlobalStatus` since each IOMMU register field consists of a single bit. And I haven't used `bitflags!` at all for `GlobalCommand` since it hasn't been necessary, although it could be added in the future if needed.

**Comment by NicholasLindsay** [comment]
> > The code style so far looks quite good! I don't see any glaring problems. I left a few small comments, nothing major. I'll let you guide the progress of this PR; feel free to reach out to me on Slack as well.
> > There's probably a better way to represent the `Capability` and `ExtendedCapability` structs, but what you've done thus far isn't necessarily wrong or bad. I'd need to see more about how those types are used in order to suggest something better, but you could always look at `bitfield` or `bitflags` for inspiration. I will say that if you do stick with this approach, it's probably clearer and more efficient to test a specific bit or a range of bits rather than right-shift the values by that bit index and then test against `0x1`, but I'm not sure if it actually compiles down to something much different.
> > My primary question is whether it's necessary to split IOMMU initialization into `early_init()` and regular `init()`. That's only really needed when we have to break cyclic dependencies, e.g., two subsystems that partially depend on each other. If not, then i'd just stick with one regular `init()` function for simplicity's sake.
> 
> Perhaps something like this would be clearer for checking bit values: `(self.0) & (1 << 63) != 0`? It's slightly more convoluted than the C version since in Rust there is no implicit conversion from `u64` to `bool`.

Stylistically I actually prefer the original version, e.g. `(self.0 >> 63) & 0x1 == 0x1`. As you said, with a good compiler, there shouldn't be much of a difference, since they are equivalent.

**Comment by NicholasLindsay** [comment]
I have done the following:
* Replace `(self.0) >> 63) & 0x1 == 0x1` style bit checks with `(self.0) & (1 << 63) != 0` which should enable more compile time optimization.
* Remove `fmt::Display` implementations for `iommu::regs::Capability` and `iommu::regs::ExtendedCapability` and add `fmt::Debug` implementations, making use of `debug_struct`.

If there are no outstanding issues then I declare this PR ready-to-go, from my point of view.

**Comment by kevinaboos** [comment]
Note to myself: this looks good, pending some minor changes that I plan to make.

**Review by kevinaboos — APPROVED** [review]

**Pull request merged**